### ISSUE TITLE
Trouble with OS X Yosemite 10.10.4

### ISSUE THREAD

**Issue by iragsdale** [droe/sslsplit#102]
I'm attempting to use sslsplit to debug IMAP conversations on my iPhone, and I'm not having much luck. Have tried with both the master branch and the develop branch. I did a standard make & make install, and sslsplit compiles with no errors, and appears to correctly support pf:
```
bash-3.2# ./sslsplit -E
pf (default)
```

I then enable port forwarding, and set up rules to divert the traffic to sslsplit:
```
 # enable port forwarding
sysctl net.inet.ip.forwarding=1
sysctl net.inet.ip.fw.enable=1
sysctl net.inet.ip.fw.verbose=1 
 
# divert traffic
echo rdr pass on en0 proto tcp from  192.168.33.241/32 to any port 143 -> 127.0.0.1 port 10143 >> ./pfrules
echo rdr pass on en0 proto tcp from  192.168.33.241/32 to any port 993 -> 127.0.0.1 port 10993 >> ./pfrules
pfctl -ef ./pfrules
```
```
bash-3.2# pfctl -sall
No ALTQ support in kernel
ALTQ related functions disabled
TRANSLATION RULES:
rdr pass on en0 inet proto tcp from 192.168.33.241 to any port = 143 -> 127.0.0.1 port 10143
rdr pass on en0 inet proto tcp from 192.168.33.241 to any port = 993 -> 127.0.0.1 port 10993
...
```
Finally I start up sslsplit to listen on those ports:
```
sslsplit -D -k ~/.mitmproxy/mitmproxy-ca.pem -l connect.log -L connectdata.log ssl 127.0.0.1 10993  tcp 127.0.0.1 10143
```

Then, I manually change the router on my phone to be the primary IP address on my Yosemite machine, and verify that traffic is flowing correctly by loading a web page, which works fine.

At this point, I'd expect that any IMAP connections I make should show up in the sslproxy logs, but they do not. I do see them attempted in pfctl:
```
STATES:
ALL tcp 127.0.0.1:10993 <- 64.233.168.108:993 <- 192.168.33.241:58096       CLOSED:SYN_SENT
ALL tcp 127.0.0.1:10993 <- 64.233.168.108:993 <- 192.168.33.241:58095       CLOSED:SYN_SENT
```
But connect.log and connectdata.log both stay at zero bytes, and I never see anything in the sslsplit debug output beyond garbage collection:
```
Started 16 connection handling threads
Starting main event loop.
Garbage collecting caches started.
Garbage collecting caches done.
``` 

I find it quite likely that this is user error, but I'm not quite sure what I could be doing wrong. If I use nc to listen on those ports, it does see the connection attempts, so it seems like sslsplit is never handling the data.

**Comment by droe** [comment]
I don't see any obvious problem with your setup.

Can you try to run nc in listening mode on 127.0.0.1:10993 to make sure that the connections redirected by pf actually end up on 127.0.0.1:10993?

Also, can you add the output of `sslsplit -V`?

**Comment by droe** [comment]
Sorry, I was not reading correctly, you actually write that you already checked listening with nc. That's weird. Output of `sslsplit -V` is still relevant. As is checking if `make test` fails any unit tests in case you haven't done so already.

**Comment by iragsdale** [comment]
Just confirmed two-way communication between a local nc -l 10993 and my device. Here's the output of sslsplit -v:

```
bash-3.2# ./sslsplit -V
SSLsplit 0.4.11-95-g3f39f58 (built 2015-07-01)
Copyright (c) 2009-2015, Daniel Roethlisberger <daniel@roe.ch>
http://www.roe.ch/SSLsplit
Build info: OSX:10.10.4 XNU:2782.1.97:fallback:2782.30.5 V:GIT
Features: -DHAVE_DARWIN_LIBPROC -DHAVE_PF
NAT engines: pf*
Local process info support: yes (Darwin libproc)
compiled against OpenSSL 0.9.8zd 8 Jan 2015 (9081df)
rtlinked against OpenSSL 0.9.8zf 19 Mar 2015 (9081df)
TLS Server Name Indication (SNI) supported
OpenSSL is thread-safe without THREADID
Not using SSL_MODE_RELEASE_BUFFERS
SSL/TLS protocol availability: ssl3 tls10 
SSL/TLS algorithm availability: RSA DSA ECDSA DH !ECDH EC
OpenSSL option availability: !SSL_OP_NO_COMPRESSION SSL_OP_NO_TICKET SSL_OP_ALLOW_UNSAFE_LEGACY_RENEGOTIATION SSL_OP_DONT_INSERT_EMPTY_FRAGMENTS SSL_OP_NO_SESSION_RESUMPTION_ON_RENEGOTIATION SSL_OP_TLS_ROLLBACK_BUG
compiled against libevent 2.0.22-stable
rtlinked against libevent 2.0.22-stable
8 CPU cores detected
```

**Comment by iragsdale** [comment]
```
test -r session.pem
./sslsplit.test
Running suite(s): 
 main
 opts
 dynbuf
 cert
 cachemgr
 cachefkcrt
 cachetgcrt
 cachedsess
 cachessess
 ssl
 sys
 base64
 url
 util
 pxythrmgr
100%: Checks: 129, Failures: 0, Errors: 0
```

**Comment by droe** [comment]
Some potential root causes, in no particular order:

-   You build against OpenSSL 0.9.8zd but run with 0.9.8zf. That should in theory be fully compatible and not an issue. But if you can, try to use the same version and installation of OpenSSL for both building and running, just in case. But see b) before you rebuild...
-   You are using a very old version of OpenSSL. You may in general get better results with a more modern version of OpenSSL that supports TLS 1.2. I recommend using the latest 1.0.2 release. It's a good idea anyway, but not terribly likely to be the culprit here.
-   Apple may have changed the PF ioctl interface in 10.10.4 in a way that causes malfunction in a way that SSLsplit does not detect. Since Apple has not released the XNU source code for 10.10.2 upwards yet, I cannot easily verify this. Not terribly likely either.
-   Some bug in libevent or SSLsplit related to the new privilege dropping code or some other changes. You may be able to test this by trying the sslsplit-0.4.11 release instead of the current `develop` branch.

**Comment by iragsdale** [comment]
I didn't realize that my installed version of OpenSSL was so old! I got the latest version of OpenSSL, and rebuilt from tag 0.4.11, and I see the same issue:
```
SSLsplit 0.4.11 (built 2015-07-01)
Copyright (c) 2009-2014, Daniel Roethlisberger <daniel@roe.ch>
http://www.roe.ch/SSLsplit
Build info: OSX:10.10.4 XNU:2782.1.97:fallback:2782.30.5 V:GIT
Features: -DHAVE_DARWIN_LIBPROC -DHAVE_PF
NAT engines: pf*
Local process info support: yes (Darwin libproc)
compiled against OpenSSL 1.0.2a 19 Mar 2015 (1000201f)
rtlinked against OpenSSL 1.0.2a 19 Mar 2015 (1000201f)
TLS Server Name Indication (SNI) supported
OpenSSL is thread-safe with THREADID
Using SSL_MODE_RELEASE_BUFFERS
SSL/TLS protocol availability: ssl3 tls10 tls11 tls12 
SSL/TLS algorithm availability: RSA DSA ECDSA DH ECDH EC
OpenSSL option availability: SSL_OP_NO_COMPRESSION SSL_OP_NO_TICKET SSL_OP_ALLOW_UNSAFE_LEGACY_RENEGOTIATION SSL_OP_DONT_INSERT_EMPTY_FRAGMENTS SSL_OP_NO_SESSION_RESUMPTION_ON_RENEGOTIATION SSL_OP_TLS_ROLLBACK_BUG
compiled against libevent 2.0.22-stable
rtlinked against libevent 2.0.22-stable
8 CPU cores detected
```

I was seeing the same issue before I upgraded from 10.10.3 to 10.10.4, but I guess they're both newer than 10.10.2, so I can't rule out the PF ioctl issue. I *can* verify that sslsplit receives traffic, as connecting to localhost:10143 generates a debug message (Connection not found in NAT state table, aborting connection).

Do you think there's any benefit to attempting to use an older version of libevent?

**Comment by droe** [comment]
That sslsplit complains about not finding the NAT state table entry when you connect to it directly (not through NAT redirection) is expected behaviour. It should not do so if it receives a connection through NAT redirection.
Next step would be to prepare an issue debugging branch of sslsplit with additional logging in the code that accepts connections and retrieves NAT state. I will not get around to prepare that today.
I would be very surprised at this point if attempting to use an older version of libevent would change anything.

**Comment by iragsdale** [comment]
Ok, I had a feeling I was doing something stupid. 

The Yosemite application firewall was the issue. I figured that out by changing the spec somewhat to listen to 0.0.0.0 instead of 127.0.0.1, at which point I got prompted to allow it to listen on the network. Once I did that, everything started working.

Alternatively I tested disabling the application firewall and using 127.0.0.1, which also seems to work OK. Might want to add that to the README for OS X users.
